File: code_snippets/java/src/main/java/guides/sagas/BookingWorkflow.java
package guides.sagas;

import dev.restate.sdk.Context;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Service;
import dev.restate.sdk.common.TerminalException;
import dev.restate.sdk.endpoint.Endpoint;
import dev.restate.sdk.http.vertx.RestateHttpServer;
import guides.sagas.clients.CarRentalClient;
import guides.sagas.clients.FlightClient;
import guides.sagas.clients.PaymentClient;
import java.util.ArrayList;
import java.util.List;

@Service
public class BookingWorkflow {

  public record BookingRequest(
      FlightClient.FlightBookingRequest flight,
      CarRentalClient.CarRentalRequest car,
      PaymentClient.PaymentInfo paymentInfo) {}

  @Handler
  public void run(Context ctx, BookingRequest req) throws TerminalException {
    var flight = req.flight;
    var paymentInfo = req.paymentInfo;
    List<Runnable> compensations = new ArrayList<>();

    // <start_twostep>
    // !mark[/bookingId/] red
    String bookingId = ctx.run(String.class, () -> FlightClient.reserve(flight));
    // !mark[/bookingId/] red
    compensations.add(() -> ctx.run(() -> FlightClient.cancel(bookingId)));

    // ... do other work, like reserving a car, etc. ...

    // !mark[/bookingId/] red
    ctx.run(() -> FlightClient.confirm(bookingId));
    // <end_twostep>

    // <start_idempotency>
    // !mark[/paymentId/] red
    String paymentId = ctx.random().nextUUID().toString();
    // !mark[/paymentId/] red
    compensations.add(() -> ctx.run(() -> PaymentClient.refund(paymentId)));
    // !mark[/paymentId/] red
    ctx.run(() -> PaymentClient.charge(paymentInfo, paymentId));
    // <end_idempotency>
  }

  public static void main(String[] args) {
    RestateHttpServer.listen(Endpoint.bind(new BookingWorkflow()));
  }
}

File: code_snippets/ts/src/guides/sagas/booking_workflow.ts
import * as restate from "@restatedev/restate-sdk";
import { flightClient, paymentClient } from "./clients/flight_client";

type BookingRequest = {
  customerId: string;
  flight: { flightId: string; passengerName: string };
  paymentInfo: string;
};

const bookingWorkflow = restate.service({
  name: "BookingWorkflow",
  handlers: {
    run: async (ctx: restate.Context, req: BookingRequest) => {
      const {customerId, flight, paymentInfo} = req;

      // create a list of undo actions
      const compensations = [];

      // <start_twostep>
      // !mark[/bookingId/] red
      const bookingId = await ctx.run(() =>
        flightClient.reserve(customerId, flight)
      );
      compensations.push(() =>
        // !mark[/bookingId/] red
        ctx.run(() => flightClient.cancel(bookingId))
      );

      // ... do other work, like reserving a car, etc. ...

      // !mark[/bookingId/] red
      await ctx.run(() => flightClient.confirm(bookingId));
      // <end_twostep>

      // <start_idempotency>
      // !mark[/paymentId/] red
      const paymentId = ctx.rand.uuidv4();
      compensations.push(() =>
        // !mark[/paymentId/] red
        ctx.run(() => paymentClient.refund(paymentId))
      );
      // !mark[/paymentId/] red
      await ctx.run(() => paymentClient.charge(paymentInfo, paymentId));
      // <end_idempotency>
    },
  },
});

restate.endpoint().bind(bookingWorkflow).listen(9080);
